text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+
+```javascript
+import { signal } from 'some/signals';
+
+export default class ExampleComponent extends LightningElement {
+    count = signal(0);
+
+    increment() {
+        this.count.value++;
+    }
+}
+```
+
+In the template, we can bind directly to the `.value` property:
+
+```html
+<template>
+    <button onclick={increment}>Increment</button>
+    <p>{count.value}</p>
+</template>
+```
+
+### Compiler Enhancements
+
+The LWC compiler will be updated to recognize when a Signal's `.value` is used within a template. Instead of accessing the property directly at runtime, it will generate code to subscribe to the Signal's changes during component's template initialization and unsubscribe upon component's template disposal.
+
+### Store Protocol
+
+A store must implement the following protocol to be compatible with the new reactivity model:
+
+```ts
+class Store {
+    get value(): any { /* return current value */ }
+    subscribe((newValue: any) => void): () => void { /* subscribe to changes */ }
+}
+```
+
+__NOTE:__ _We can work the details of this protocol. It could be more simple, or hopefully similar to what is out there._
+
+### Integration with Components
+
+Components should treat signals as internal reactive state mechanisms. While a signal can technically be passed down to child components, it's recommended to pass only the necessary data, typically the `.value` of the signal, to keep the child components decoupled from the parent's state management strategy.
+
+Here’s how a parent component would pass the current value of a signal to a child component, which expects a number:
+
+```javascript
+// ParentComponent.js
+import { signal } from 'some/signals';
+
+export default class ParentComponent extends LightningElement {
+    count = signal(0);
+}
+```
+
+```html
+<!-- ParentComponent's template -->
+<template>
+    <!-- Pass the value of the signal, not the signal itself -->
+    <child-component count={count.value}></child-component>
+</template>
+```
+
+The child component simply receives a number as a property and uses it as such, unaware of the signal mechanism in the parent:
+
+```javascript
+export default class ChildComponent extends LightningElement {
+    @api count;
+}
+```
+
+This ensures that child components remain agnostic of the parent's state management implementation, promoting better component encapsulation and reuse.
+
+In terms of optimizations, the compiler can detect when a signal is passed down to a child component because of the `.value` in the template interpolation, and generate code to subscribe to the signal's changes. This would allow the parent component to update the child's public property `count` without a full re-render, leading to better performance. On the receiving end, because the child component's `count` property is not a signal, but a value, it will react to changes as usual, without any special handling.
```

Comment:
What is the use case for being able to pass the signal ref instead of `.value` to a child component? Again, as a dev I really did not like to use `count.value`. If there is no use case, I think we could just use `count`. Then the framework can identify that it is a signal based on its type (maybe you could officially make lwc ts first to get errors at build time) and that the child has to react based on `.value`

Reply:
there are 3 ways to provide stores to a component:

1. create or import a signal object
2. use a global signal (not recommended)
3. get it via a prop

So take your poison. If you have two side by side components that are using the same store provisioned by a parent component, what else can you do other than 3?